2022-RADIUSS-AWS/JupyterNotebook/radiuss-aws-flux.ipynb

```diff
@@ -0,0 +1,944 @@
+   "metadata": {},
+   "source": [
+    "# Flux Process and Job Utilities\n",
+    "## Flux top\n",
+    "Flux provides a feature-full version of `top` for nested Flux instances and jobs. In the JupyterLab terminal, invoke `flux top` to see the \"sleep\" jobs. If they have already completed you can resubmit them. \n",
+    "\n",
+    "We recommend not running `flux top` in the notebook as it is not designed to display output from a command that runs continuously.\n",
+    "\n",
+    "## Flux pstree\n",
+    "In analogy to `top`, Flux provides `flux pstree`. Try it out in the JupyterLab terminal or here in the notebook.\n",
+    "\n",
+    "## Flux proxy\n",
+    "### Interacting with a job hierarchy with `flux proxy`\n",
+    "#### Example 1\n",
+    "Routes messages to and from a Flux instance. We can use `flux proxy` to connect to a running Flux instance and then submit more nested jobs inside it. You may want to edit `sleep_batch.sh` with the JupyterLab text editor (double click the file in the window on the left) to sleep for `60` or `120` seconds. Then from the JupyterLab terminal, run: \n",
+    "\n",
+    "`flux mini batch --nslots=2 --cores-per-slot=1 --nodes=2 ./sleep_batch.sh` # outputs JOBID\n",
+    "\n",
+    "`flux pstree -x`\n",
+    "\n",
+    "`flux proxy JOBID` # this connects you to the Flux instance corresponding to JOBID above\n",
+    "\n",
+    "`flux uptime` # Note the depth is now 1 and the size is 2: we're one level deeper in a Flux hierarchy and we have only 2 brokers now.\n",
+    "\n",
+    "`flux resource list` # This instance has 2 \"nodes\" and 2 cores allocated to it\n",
+    "\n",
+    "`flux top`\n",
+    "\n",
+    "#### Example 2\n",
+    "Contents of `sleeps.sh`:\n",
+    "\n",
+    "``` bash \n",
+    "#!/bin/bash\n",
+    "\n",
+    "flux mini submit -N1 sleep 30\n",
+    "flux mini submit -N1 sleep 30\n",
+    "flux mini submit -N1 sleep 30\n",
+    "flux mini submit -N1 sleep 30\n",
+    "flux queue drain\n",
+    "```\n",
+    "Note `flux queue drain` which waits for the sub-jobs to complete.\n",
+    "`flux mini batch -N2 sleeps.sh`\n",
+    "(copy output JOBID)\n",
+    "`flux proxy JOBID`\n",
+    "`flux jobs`\n",
+    "#### Example 3\n",
+    "Here's an example of submitting jubs within a nested instance. You can run this example here in the notebook."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "2e6976f8-dbb6-405e-a06b-47c571aa1cdf",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!cat sub_job1.sh\n",
+    "!cat sub_job2.sh"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "8640a611-38e4-42b1-a913-89e0c76c8014",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!flux mini batch -N1 ./sub_job1.sh"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "2d2b1f0b-e6c2-4583-8068-7c76fa341884",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!flux pstree"
```

Comment:
Also show `flux pstree -x` to show extended details of jobs.

Reply:
That example is on line 516.